File: 1. two sum.c
//One-pass Hash Table

int* twoSum(int* nums, int n, int target, int* returnSize) {
    struct numMap {
        int key;
        int value;
        UT_hash_handle hh;
    } *numMap = NULL, *item;
    

    for (int i = 0; i < n; i++) {
        int complement = target - nums[i];
        HASH_FIND_INT(numMap, &complement, item);
        if (item) {
            int* result = malloc(sizeof(int) * 2);
            result[0] = item->value;
            result[1] = i;
            *returnSize = 2;
            HASH_CLEAR(hh, numMap);  // Free the hash table
            return result;
        }
        item = malloc(sizeof(struct numMap));
        item->key = nums[i];
        item->value = i;
        HASH_ADD_INT(numMap, key, item);
    }
    *returnSize = 0;
    HASH_CLEAR(hh, numMap);  // Free the hash table
    // Return an empty array if no solution is found
    return malloc(0);  // Allocate 0 bytes
}

File: 1. two sum.py
#One-pass Hash Table
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        numMap = {}   #建立空的雜湊表
        n = len(nums) #獲取陣列的長度，方便後續迴圈使用

        for i in range(n):   #開始遍歷陣列 nums , i 代表當前元素的索引
        
            complement = target - nums[i]     #計算我們需要的「另一半」數字

            if complement in numMap:   #在雜湊表查找
                
                #如果找到，直接回傳兩個數字的索引
                return [numMap[complement],i]  

            numMap[nums[i]] = i      ##如果沒找到，將當前數字及其索引存入雜湊表

        return []   #no solution found
